Notify and prevent user from creating application using unsupported GUI frameworks for mobile platforms

qlchan24 opened the pull request:
## Changes:
<!--- Describe your changes in detail -->
* Register and check for the new `supported` attribute of `pyproject.toml` for `AppConfig` in `config.py`
* Implement `GuiUnsupportedForPlatform` exception for the validation below.
* Add validation in `CreateCommand#create_app()` and `UpdateCommand#update_app()` which throws the `GuiUnsupportedForPlatform` exception if the GUI framework is not supported for the platform chosen.
* Edit displayed string for the `new` command wizard in `build_app_context()` of `new.py` to notify users of the unsupported platforms for the GUI selection.
* Update context in `test_generate_app_template.py` to include the new metadata attribute
* Add documentation for new configuration option in `configuration.rst`
* Add note to inform users of the unsupported GUI frameworks in the documentation for `android.rst` and `iOS.rst`
* Correct typo in the comment for `parse_config()` in `config.py`

## Problem Solved:
<!--- What problem does this change solve? -->
Briefcase does not support deployment of GUI frameworks Pyside2 and PPB to iOS or Android, even though they are valid options in the `new` command. This pull will help to remind the user when using the `new` command wizard, and disallow the user to deploy unsupported GUI frameworks when using the `create` or `update` commands.

More details provided in the related issue below.

**Note that this requires the [template change](https://github.com/beeware/briefcase-template/pull/24) from the briefcase-template repository.**

## Related Issues:
<!--- If this PR relates to an issue, include Refs #XXX or Fixes #XXX -->
Briefcase issue #539: [Provide better warnings when an app can't be deployed to mobile platforms](https://github.com/beeware/briefcase/issues/539)

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

freakboy3742 reviewed (changes requested):
Looking good! I've flagged a couple of items inline - they're mostly cosmetic.

The one piece that *is* missing (other than the flake8 formatting failure) is tests. You've added a non-trivial piece of functionality, so we need additional tests to make sure that a configuration that has `supported=False` will be handled correctly.

qlchan24 commented on src/briefcase/config.py:
```diff
@@ -216,7 +218,7 @@ def merge_config(config, data):
     :param data: The new configuration data to merge into the configuration.
     """
-    for option in ['requires', 'sources']:
+    for option in ['requires', 'sources', 'is_supported']:
```
Yes, it should be `supported`, I'll fix that!

qlchan24 replied:
Actually, I will remove the `supported` element from that list as this would result in the `supported` field being `true` even with the `supported=false` attribute in the platform specific toml attributes. Removing it altogether results in the intended outcome.

qlchan24 commented on src/briefcase/config.py:
```diff
@@ -120,6 +120,7 @@ def __init__(
         splash=None,
         document_type=None,
         template=None,
+        supported=None,
```
True, I will change this too

qlchan24 commented on docs/reference/platforms/android.rst:
```diff
@@ -2,7 +2,8 @@
 =======
 
-When generating an Android project, Briefcase produces a Gradle project.
+When generating an Android project, Briefcase produces a Gradle project. Note that the GUI frameworks PySide2 and
```
Ok, I'll delete this and the one for ios then.

qlchan24 commented on src/briefcase/commands/new.py:
```diff
@@ -420,7 +420,7 @@ def build_app_context(self):
         gui_framework = self.input_select(
             intro="""
-What GUI toolkit do you want to use for this project?""",
+What GUI toolkit do you want to use for this project? Note: PySide2 and PursuedPyBear cannot be deployed to Android or iOS""",
```
Ok, I'll use that for the options. As for how it would be processed, I'm thinking of just extracting out the corresponding string. eg: 'Pyside2' from the string 'PySide2 (does not support iOS/Android deployment)'. Or should I update the names used in the briefcase template instead?

qlchan24 commented:
I have fixed the mentioned problems and I have also implemented tests to cover the validation code in the `create` and `update` commands and the parsing of the GUI options in the `new` command.

freakboy3742 commented on changes/544.feature.rst:
```diff
@@ -0,0 +1,2 @@
+Modified GUI options in new project wizard to advise on platform compatibility,
+and disallow deployment on incompatible platforms through the create and update commands.
```
This should be "539.feature.rst", not 544, so that the issue number tracks the original feature request that is being closed.

freakboy3742 reviewed (changes requested):
This is starting to look really good. 

The one thing that remains is some tweaking to the docs and release notes. Ordinarily, I'd do that myself, as it's usually easier for me to make minor language/grammar changes. However, you've submitted this PR against your master branch, which means I don't have permission to commit to the PR. 

Could I get you to re-submit this PR using a branch? Submitting PRs from a branch is generally a good idea regardless - as well as allowing project maintainers to tweak PRs, it makes it a lot easier to keep your fork of a project up to date with the origin project.

You should be able to create a branch on Github by pushing your current master as a new branch name:

    git push origin master:fix-539

and submitting a pull request from that new branch. (I picked the name `fix-539` as an identifier because this PR is a fix for issue 539 - but you could pick any name you want)

qlchan24 commented:
Yes my apologies for submitting from master, I have created the new branch and PR as suggested, with the new changes, at #546.

freakboy3742 commented:
Closing in favour of the non-master version in #546
